src/modules/endpoint/sagas/watchFetchNews.js

```diff
@@ -0,0 +1,213 @@
+  FetchMoreNewsActionType,
+  FetchLanguagesFailedActionType
+} from '../../app/StoreActionType'
+import type { DataContainer } from '../DataContainer'
+import { ContentLoadCriterion } from '../ContentLoadCriterion'
+import isPeekingRoute from '../selectors/isPeekingRoute'
+import ErrorCodes, { fromError } from '../../error/ErrorCodes'
+import { LOCAL } from '../../../routes/news/containers/WithCustomNewsProvider'
+import loadLocalNews from './loadLocalNews'
+import loadTunews from './loadTunews'
+import loadLanguages from './loadLanguages'
+import loadTunewsElement from './loadTunewsElement'
+
+const TUNEWS_FETCH_COUNT_LIMIT = 20
+const FIRST_PAGE_INDEX = 1
+
+export function * fetchNews (
+  dataContainer: DataContainer,
+  action: FetchNewsActionType
+): Saga<void> {
+  const { city, language, newsId, key, type } = action.params
+  try {
+    const isLocalNews = type === LOCAL
+    // Update language each time when switching between news
+    const languages = yield call(
+      loadLanguages,
+      city,
+      dataContainer,
+      true,
+      !isLocalNews
+    )
+    const languageValid = languages
+      .find(languageModel => languageModel.code === language)
+    const pushLanguagesAction = {
+      type: 'PUSH_LANGUAGES',
+      params: { languages: languages }
+    }
+    yield put(pushLanguagesAction)
+
+    if (languageValid) {
+      const [news] = yield all([
+        isLocalNews
+          ? call(loadLocalNews, city, language)
+          : newsId
+            ? yield call(loadTunewsElement, parseInt(newsId, 0)) // A better solution to prevent re-fetching news again from page 1
+            : call(
+              loadTunews,
+              language,
+              FIRST_PAGE_INDEX,
+              TUNEWS_FETCH_COUNT_LIMIT
+            )
+      ])
+
+      const insert: PushNewsActionType = {
+        type: 'PUSH_NEWS',
+        params: {
+          news,
+          newsId,
+          cityLanguages: languages,
+          key,
+          language,
+          city,
+          type,
+          page: FIRST_PAGE_INDEX,
+          hasMoreNews: true
+        }
+      }
+      yield put(insert)
+    } else {
+      const allAvailableLanguages = newsId ? new Map(languages.map(language => [language.code, null])) : null
+
+      const failed: FetchNewsFailedActionType = {
+        type: 'FETCH_NEWS_FAILED',
+        params: {
+          message: 'Could not load news.',
+          code: ErrorCodes.PageNotFound,
+          allAvailableLanguages,
+          newsId: null,
+          key,
+          type,
+          language,
+          city
+        }
+      }
+      yield put(failed)
+    }
+  } catch (e) {
+    console.error(e)
+    const failed: FetchNewsFailedActionType = {
+      type: 'FETCH_NEWS_FAILED',
+      params: {
+        message: `Error in fetch news: ${e.message}`,
+        code: fromError(e),
+        key,
+        city,
+        language,
+        type,
+        newsId,
+        allAvailableLanguages: null
```

Comment:
I think this is not correct according to the flow types. Maybe an empty map would be better here.

Reply:
Ok

Reply:
If there are other occurrences in our code than this should be fine. Didn't check that though.